### ISSUE TITLE
[REACT] custom eventRenderer which return JSX, cause a significant delay in rendering events

### ISSUE THREAD

**Issue by kronaemmanuel** [bryntum/support#9370]
I've done initial investigation along with discussion with Saki. We are doing a lot of re-renders, however it's not clear where we can shave off the re-renders. Here's the Slack thread: https://bryntum.slack.com/archives/G011DNLMM3Q/p1718173531730379

Idea is to try to not call `updateGeneration` as much but this could lead to artifacts in rendering.

Notes:
- This problem is not due to `flushSync` itself, but due to how many times `updateGeneration` is called

[Forum post](https://forum.bryntum.com/viewtopic.php?f=51&t=29354&p=148245#p148245)

When using a custom eventRenderer that returns JSX, there is a significant delay in rendering events when scrolling, compared to when it returns a string. Additionally, multiple warnings appear in the console:

```
Warning: flushSync was called from inside a lifecycle method. React cannot flush when React is already rendering. Consider moving this call to a scheduler task or micro task.
```

I want to attach a minimal reproducible example to this post, but there is an error when attaching it:

```
HTTP Error.
```
and in console:

```
Failed to load resource: net::ERR_CONNECTION_TIMED_OUT
```

**Comment by marciogurka** [comment]
[+1](https://forum.bryntum.com/viewtopic.php?p=148599#p148599) for this one

**Comment by rheutz91** [comment]
+1

**Comment by mykels** [comment]
+1

**Comment by jsakalos** [comment]
The complain that states that "strings perform better than React components" are all true and will always be true. When React is used, a level of complexity and overhead is added. The React components must be create and instantiated as portals at some point and then destroyed at another point. Therefore, native strings or DomConfig returned from render WILL ALLWAYS BE FASTER than React renderers. 

Therefore, when we need hundreds or thousands or even more event or records simultaneously existing in DOM we must make them as simple as possible stripping down any possible overheads – of which React takes a significant part.

Another approach could be to not need that many DOM elements and thus React components at the same time. This can be achieved by efficient and user-friendly searching and filtering. In other words, we would rather have user to find his things than to scroll (infinitely) to his things. Humans cannot perceive thousands of elements at the same time anyway; perhaps 10-20.

At the dawn of computers, but I believe also nowadays, if a highly performant, real-time app is necessary then it is written as close to machine code as possible - perhaps still in assembler.

**Comment by mykels** [comment]
Hi @jsakalos,

Thank you for your response. I appreciate your team's efforts in maintaining the Bryntum Scheduler Pro.

I've conducted a thorough profiling on the Bryntum Scheduler Pro component and I've noticed a significant performance degradation since the introduction of `flushSync` usage. This issue seems to occur only when rendering as a JSX component for the event.

In this case, it seems that `flushSync` was introduced to solve the flickering issue. While it might have solved that problem, it introduced another one - slowness in various scheduler interactions like scroll, zoom, etc. This is likely due to the blocking of the main thread and inefficient rendering caused by the usage of the `flushSync` API.

I've attached a video and my profiler project demonstrating the differences. I believe these findings will provide valuable insights into the performance issues we're experiencing.

I understand that the warning is currently considered non-critical, but I believe addressing this performance issue could greatly improve the user experience for us and other users of the Bryntum Scheduler Pro.

I look forward to your thoughts on this matter @matsbryntse 

Video

https://github.com/bryntum/support/assets/6383494/13ded33c-dd33-4ddd-9263-7c92aa7f9d46

Profiler Project
[acs-bryntum-profiler.zip](https://github.com/user-attachments/files/16049763/acs-bryntum-profiler.zip)

Best regards,
Micha

**Comment by matsbryntse** [comment]
Thanks for your feedback. Our rendering is synchronous, as we need to update the row height after all events have been laid out. Did you look deeper into this where the CPU cycles are consumed? Is it in your JSX code or in React core?

**Comment by jsakalos** [comment]
Thank you for your code Micha. I'll make some tests with the aim of making `flushSync` configurable. Will let you know when I'll have some results.

**Comment by mykels** [comment]
> Thanks for your feedback. Our rendering is synchronous, as we need to update the row height after all events have been laid out. Did you look deeper into this where the CPU cycles are consumed? Is it in your JSX code or in React core?

I have looked deeper into the CPU usage. To minimize the impact of the JSX code, I rendered a simple <div/> element, but the performance still degraded between versions 5.6.6 and 5.6.8, where the flushSync API was introduced.

Making the use of flushSync configurable would be ideal for us @jsakalos, as we calculate the row height ourselves in our actual source code (not in the attached profiler code). If there’s a way to avoid using flushSync without reintroducing the flickering issue, that would be **fantastic**.

**Comment by jsakalos** [comment]
Hello Michael, if we made `flushSync` configurable then it would be either-or proposition so it would either be used (no flickering + performance impact) or not (flickering + no performance impact).

Anyway, I'm thinking about putting some intelligence there. We could perhaps sense that the user is dragging so `flushSync` would kick-in to prevent flickering but it would be off in other situation - this approach could give us the best of both worlds.

One more thing to consider: We have released the new major version 6.0.0 recently so upgrade please your showcase. Generally, we do not support older release so we won't backport any solution to the 5.x series.

**Comment by jsakalos** [comment]
I have made some changes in the wrapper. Michael, would you please test with the attached npm?
[bryntum-schedulerpro-react-6.0.1.tgz](https://github.com/user-attachments/files/16069759/bryntum-schedulerpro-react-6.0.1.tgz)

**Comment by mykels** [comment]
Thanks @jsakalos I will test it as soon as I can and update regarding the findings. We're perfectly fine with you guys pushing this fix only to the latest major, it's understandable.

**Comment by mykels** [comment]
Hey,

I'm trying to use this but I'm getting the following error when running the application:

```
TypeError: Cannot destructure property 'assignmentRecord' of '(intermediate value)(intermediate value)(intermediate value)' as it is undefined.
    at Object.processEventContent (WrapperHelper.tsx:681:11)
```

I didn't see any breaking changes regarding the assignment model so I'm not sure what is the source of this issue. Can you please assist?

I'm attaching my profiler code with the latest bryntum version and your `6.0.1` source code.

[acs-bryntum-profiler_6.1_hotfix.zip](https://github.com/user-attachments/files/16072538/acs-bryntum-profiler_6.1_hotfix.zip)

**Comment by jsakalos** [comment]
Hello Michael,

although we still need to investigate the true reason for that error, I've made an emergency fix that would let you continue and test the performance improvements. Please use the following npm for that:

[bryntum-schedulerpro-react-6.0.1.tgz](https://github.com/user-attachments/files/16080925/bryntum-schedulerpro-react-6.0.1.tgz)

**Comment by mykels** [comment]
Hey Saki,

I was able to run the profiler using your second fix. It actually looks much better!

You can see in the video that:

1. During the performance test no `flushSync` errors are thrown.
2. The recorded performance test times are 200-300% better than the video above demonstrating the `5.6.8` version experience.

Some of the improvements may be attributed to the latest major version itself which I can not test yet since it lacks the other fix you've made. In any case - it's definitely good news for us.
 
https://github.com/bryntum/support/assets/6383494/e4fb7253-9558-4bde-80c9-1e9bdc8c8dbc

**Comment by jsakalos** [comment]
Super Michael,

glad to hear that. I hope you can continue now. Also, I cannot guarantee that this fix will be released in 6.0.1 version so continue with what I've given you until publicly released.

FYI, `flushSync` is now only called after D&D operations, not on scroll or zoom.

**Comment by jsakalos** [comment]
Reopening the issue because we have a fix now.

**Comment by shermanmicha** [comment]
Hey @jsakalos, any estimation when is this change going to be merged?

**Comment by matsbryntse** [comment]
It was released in 6.0.2

**Comment by shermanmicha** [comment]
> It was released in 6.0.2

Oh we haven't seen it in the CHANGELOG sorry. We'll check that out.
